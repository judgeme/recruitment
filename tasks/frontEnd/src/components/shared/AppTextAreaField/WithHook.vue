<script setup lang="ts">
import { defineProps, withDefaults, defineOptions } from 'vue';
import { useField } from 'vee-validate';
import AppTextAreaField, { type Props } from '@/components/shared/AppTextAreaField/index.vue';

interface WithHooksProps extends Props {
  name: string; // name is required here
  required?: boolean;
}

defineOptions({
  name: 'AppTextAreaFieldWithHook' // set up display name in dev-tools
});

const props = withDefaults(defineProps<WithHooksProps>(), {});

const { value, errorMessage, handleBlur, handleChange } = useField<string>(
  () => props.name,
  undefined,
  {
    syncVModel: true
  }
);
</script>
<template>
  <app-text-area-field
    :label="label"
    :helper-text="errorMessage"
    :invalid="!!errorMessage"
    :name="name"
    :model-value="value"
    :id="id"
    :required="required"
    :disabled="disabled"
    @update:model-value="handleChange"
    @blur="handleBlur"
  />
</template>
