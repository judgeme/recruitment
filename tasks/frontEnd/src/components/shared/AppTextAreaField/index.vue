<script setup lang="ts">
import { defineProps, withDefaults, defineModel, defineOptions, defineEmits } from 'vue';
import type { TextareaHTMLAttributes } from 'vue';

import AppFormItem from '@/components/shared/AppFormItem/index.vue';

// extends InputHTMLAttributes to have access to native input props
export interface Props extends /* @vue-ignore */ TextareaHTMLAttributes {
  id?: string;
  name?: string;
  loading?: boolean;
  disabled?: boolean;
  label?: string;
  required?: boolean;
  invalid?: boolean;
  helperText?: string;
  autoComplete?: string;
}

withDefaults(defineProps<Props>(), {
  autoComplete: 'off'
});

defineOptions({
  name: 'AppTextAreaField', // set up display name in dev-tools
  inheritAttrs: false
});

defineEmits<{
  blur: [value: FocusEvent];
  'update:model-value': [value: string];
}>();

const modelValue = defineModel<string>();
</script>
<template>
  <app-form-item
    :htmlFor="id"
    :label="label"
    :helper-text="helperText"
    :invalid="invalid"
    :required="required"
  >
    <textarea
      v-model="modelValue"
      v-bind="$attrs"
      class="
        appearance-none block w-full 
        py-[15px] px-5 font-inherit font-light 
        text-input text-primary bg-white 
        border outline-0 transition-colors 
        focus:bg-fairy placeholder:text-grey-40
        disabled:cursor-not-allowed
      "
      :class="invalid ? 'border-danger' : 'border-grey-60'"
      :id="id"
      :required="required"
      :name="name"
      :disabled="disabled"
      :autocomplete="autoComplete"
      @blur="$emit('blur', $event)"
    />
  </app-form-item>
</template>
