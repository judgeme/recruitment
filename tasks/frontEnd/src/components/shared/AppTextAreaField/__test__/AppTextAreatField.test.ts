import { mount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';
import AppTextAreaField from '@/components/shared/AppTextAreaField/index.vue';

describe('AppTextAreaField.vue', () => {
  it('renders textarea element with default props', () => {
    const wrapper = mount(AppTextAreaField);

    // Check if the textarea element is rendered
    const textarea = wrapper.find('textarea');
    expect(textarea.exists()).toBe(true);

    // Check if default classes are applied
    expect(textarea.classes()).toContain('block');
    expect(textarea.classes()).toContain('w-full');
    expect(textarea.classes()).toContain('py-[15px]');
    expect(textarea.classes()).toContain('px-5');
    expect(textarea.classes()).toContain('text-primary');
    expect(textarea.classes()).toContain('bg-white');
    expect(textarea.classes()).toContain('border');
    expect(textarea.classes()).toContain('outline-0');
    expect(textarea.classes()).toContain('transition-colors');
    expect(textarea.classes()).toContain('placeholder:text-grey-deep');

    // Check if default attributes are applied
    expect(textarea.attributes('autocomplete')).toBe('off');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('applies custom props and classes correctly', async () => {
    const wrapper = mount(AppTextAreaField, {
      props: {
        id: 'test-id',
        name: 'test-name',
        label: 'Test Label',
        type: 'email',
        required: true,
        invalid: true,
        helperText: 'Helper Text',
        disabled: true,
      },
    });

    const textarea = wrapper.find('textarea');
    expect(textarea.attributes('id')).toBe('test-id');
    expect(textarea.attributes('name')).toBe('test-name');
    expect(textarea.attributes('autocomplete')).toBe('off');
    expect(textarea.classes()).toContain('border-danger');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('emits blur event on textarea blur', async () => {
    const wrapper = mount(AppTextAreaField);
    const textarea = wrapper.find('textarea');

    await textarea.trigger('blur');

    expect(wrapper.emitted('blur')).toBeTruthy();
  });

  it('emits model value event on textarea changes', async () => {
    const wrapper = mount(AppTextAreaField);
    const textarea = wrapper.find('textarea');

    await textarea.trigger('input');

    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
  });

});
