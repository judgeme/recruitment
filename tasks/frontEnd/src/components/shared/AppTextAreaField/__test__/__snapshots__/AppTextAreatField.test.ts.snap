// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppTextAreaField.vue > applies custom props and classes correctly 1`] = `
"<div class="mb-2.5 w-full">
  <div class="relative"><label class="block tracking-wide text-primary text-anchor font-medium mt-0 mb-1" for="test-id"><span class="text-ellipsis">Test Label</span><span class="whitespace-nowrap ml-1">*</span></label><textarea type="email" class="appearance-none block w-full py-[15px] px-5 font-inherit font-light text-input text-primary bg-white border outline-0 transition-colors focus:bg-fairy placeholder:text-grey-40 disabled:cursor-not-allowed border-danger" id="test-id" required="" name="test-name" disabled="" autocomplete="off"></textarea></div>
  <div role="alert" class="text-xs mt-0.5 text-danger italic">Helper Text</div>
</div>"
`;

exports[`AppTextAreaField.vue > renders textarea element with default props 1`] = `
"<div class="mb-2.5 w-full">
  <div class="relative">
    <!--v-if--><textarea class="appearance-none block w-full py-[15px] px-5 font-inherit font-light text-input text-primary bg-white border outline-0 transition-colors focus:bg-fairy placeholder:text-grey-40 disabled:cursor-not-allowed border-grey-60" autocomplete="off"></textarea>
  </div>
  <!--v-if-->
</div>"
`;
